server/market/types/types.go

```diff
@@ -0,0 +1,167 @@
+		LotSize: lotSize,
+	}, nil
+}
+
+// OrderStatus indicates the state of an order.
+type OrderStatus uint16
+
+// There are two general classes of orders: ACTIVE and ARCHIVED. Orders with one
+// of the ACTIVE order statuses that follow are likely to be updated.
+const (
+	// OrderStatusUnknown is a sentinel value to be used when the status of an
+	// order cannot be determined.
+	OrderStatusUnknown OrderStatus = iota
+
+	// OrderStatusEpoch is for orders that have been received and validated, but
+	// not processed yet by the epoch order matcher.
+	OrderStatusEpoch
+
+	// OrderStatusBooked is for orders that have been put on the book
+	// ("standing" time in force). This includes partially filled orders. As
+	// such, when an order with this "booked" status is matched with another
+	// order, it should have its filled amount updated, and its status should
+	// only be changed to OrderStatusExecuted if the remaining quantity becomes
+	// less than the lot size, or perhaps to OrderStatusFailed if the swap has
+	// failed and DEX policy prevents it from re-entering the order book.
+	OrderStatusBooked
+
+	// OrderStatusExecuted is for orders that have been successfully processed
+	// and taken off the book. An order may reach this state if it is (1)
+	// matched one or more times and removed from the books, or (2) unmatched in
+	// epoch processing and with a time-in-force that forbids the order from
+	// entering the books. Orders in the first category (matched and
+	// subsequently removed from the book) include: a matched cancel order, a
+	// completely filled limit or market order, or a partially filled market buy
+	// order. Market and limit orders in the second category necessarily will
+	// necessarily be completely unfilled. Partially filled orders that are
+	// still on the order book remain in OrderStatusBooked.
+	//
+	// QUESTION: How do we distinguish a CancelOrder that was not matched from
+	// one that was matched? With a limit or market order that was not matched,
+	// the filled amount shows this, but cancel orders have no meaningful filled
+	// amount. Having to search a matches table is undesirable, especially since
+	// it is a worst-case query returning zero rows.
+	OrderStatusExecuted
+
+	// OrderStatusCanceled is for orders that were on the book, but matched with
+	// a cancel order. This does not mean the order is completely unfilled.
+	OrderStatusCanceled
+
+	// QUESTION: What status is for standing limit orders that were matched, but
+	// have failed to swap, and do not re-enter the book? Are they forcibly
+	// canceled (OrderStatusCanceled) or are they considered executed
+	// (OrderStatusExecuted)? Does this also depend on DEX policy regarding the
+	// circumstances of swap failure?
+
+	// The above QUESTIONs suggest that a failed status may be needed, although
+	// perhaps just internal to the DB backend's implementation, to quickly
+	// identify orders that failed to match or failed in the swap stage.
+)
```

Comment:
Following the conversation at https://github.com/decred/dcrdex/pull/47#discussion_r334691126, the statuses were reworked.   @buck54321, how would you address the `QUESTION`s noted above?

Having the two terminal states be "canceled" and "executed" only serves to identify orders that were matched with a cancel order from those that were not.  One peculiarity is that the only possible terminal state for a cancel order is "executed", and to determine if it actually matched or not requires querying a matches table (a negative result is a a more time consuming query).  With limit and market orders, the quantity and filled amounts indicate how it got to executed status (zero vs 1+ matches), but we don't get that with a cancel order without referencing another table.

Identifying orders that failed during swap stage via the swaps table is more tolerable, but it is still not ideal because executed orders with a 0 filled amount still may have a (failed) entry in the swaps table.

Reply:
**Documenting off-github discussion.** Hopefully I captured the important parts. 

We discussed having more independence between matches and orders. There was some question of what the `Filled` field on an order represents: the sum of match quantities, or the successfully swapped quantity. My thinking is towards the latter, with orders able to be in an executed or canceled state even if some swaps are active. In the case of failed swaps, clients can accommodate auto-reorder through configuration or UI, IMO.

I'm not 100% certain where we ended on the question of how to specify whether a cancel order matched or not, but there are some options. If you need to add an additional status or two, you should, though. I think we should also differentiate between client-cancelled orders, `OrderStatusCanceled`, and DEX-revoked orders `OrderStatusRevoked`. 

Regardless, we don't have to get this all pinned down now, This will all come together once we start integrating all the parts, so probably important not to get hung up on anything here. We also have some freedom to play with the DB without writing DB upgrade code for a while, which is nice.

Reply:
I though I had responded to this comment, sorry.

Regarding is a cancel order matched or not, there is an internal status for the cancel table `orderStatusFailed` which gets set for a cancel order via the `FailCancelOrder(co *order.CancelOrder) error` function, which is part of the reworked `OrderArchiver` `interface`.

Regarding differentiating between client-canceled and DEX-revoked orders, I can do two things together that will support this:
- Add a new exported status `OrderStatusRevoked` for use with `UpdateOrderStatus`.
- Add a new function to `OrderArchiver` called `RevokeOrder(*order.LimitOrder) error` that would complement `CancelOrder(*order.LimitOrder) error`

Reply:
Regarding the semantics of order table entries, the Filled field, and swaps, we agree:
- Orders may be executed or canceled (no more updates) even though the swaps are active.
- Failed swaps should **not** resuscitate an order.  Rather, the client must handle reorder.

This means that the Filled field represents **the sum of match quantities NOT successfully swapped quantity.**

Reply:
Revoked status stuff in 600706df3e78711de8ef340c4a2d317ed97da34d.